// angular-plotly.js/fesm2022/angular-plotly.js.mjs
import * as i0 from '@angular/core';
import { Injectable, input, output, ViewChild, Component, NgModule } from '@angular/core';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';

class PlotlyService {
    static { this.instances = []; }
    static { this.plotly = undefined; }
    static { this.moduleName = undefined; }
    static setModuleName(moduleName) {
        PlotlyService.moduleName = moduleName;
    }
    static getModuleName() {
        return PlotlyService.moduleName;
    }
    static setPlotly(plotly) {
        if (typeof plotly === 'object' && typeof plotly.react !== 'function') {
            throw new Error('Invalid plotly.js version. Please, use any version above 1.40.0');
        }
        PlotlyService.moduleName = 'PlotlyJS';
        PlotlyService.plotly = plotly;
    }
    static insert(instance) {
        const index = PlotlyService.instances.indexOf(instance);
        if (index === -1) {
            PlotlyService.instances.push(instance);
        }
        return instance;
    }
    static remove(div) {
        const index = PlotlyService.instances.indexOf(div);
        if (index >= 0) {
            PlotlyService.instances.splice(index, 1);
            PlotlyService.plotly.purge(div);
        }
    }
    getInstanceByDivId(id) {
        for (const instance of PlotlyService.instances) {
            if (instance && instance.id === id) {
                return instance;
            }
        }
        return undefined;
    }
    async getPlotly() {
        await this.waitFor(() => this._getPlotly() !== 'waiting');
        return this._getPlotly();
    }
    _getPlotly() {
        if (typeof PlotlyService.plotly === 'undefined') {
            const msg = PlotlyService.moduleName === 'ViaCDN'
                ? `Error loading Peer dependency plotly.js from CDN url`
                : `Peer dependency plotly.js isn't installed`;
            throw new Error(msg);
        }
        return PlotlyService.plotly;
    }
    waitFor(fn) {
        return new Promise((resolve) => {
            const localFn = () => {
                fn() ? resolve() : setTimeout(localFn, 10);
            };
            localFn();
        });
    }
    async newPlot(div, data, layout, config, frames) {
        await this.waitFor(() => this._getPlotly() !== 'waiting');
        if (frames) {
            const obj = { data, layout, config, frames };
            return this._getPlotly().newPlot(div, obj).then(() => PlotlyService.insert(div));
        }
        return this._getPlotly().newPlot(div, data, layout, config).then(() => PlotlyService.insert(div));
    }
    plot(div, data, layout, config, frames) {
        if (frames) {
            const obj = { data, layout, config, frames };
            return this._getPlotly().newPlot(div, obj);
        }
        return this._getPlotly().newPlot(div, data, layout, config);
    }
    update(div, data, layout, config, frames) {
        if (frames) {
            const obj = { data, layout, config, frames };
            return this._getPlotly().react(div, obj);
        }
        return this._getPlotly().react(div, data, layout, config);
    }
    resize(div) {
        return this._getPlotly().Plots.resize(div);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });

/* eslint-disable @angular-eslint/no-conflicting-lifecycle */
// @dynamic
class PlotlyComponent {
    constructor(plotly, iterableDiffers, keyValueDiffers) {
        this.plotly = plotly;
        this.iterableDiffers = iterableDiffers;
        this.keyValueDiffers = keyValueDiffers;
        this.defaultClassName = 'js-plotly-plot';
        this.data = input();
        this.layout = input();
        this.config = input();
        this.frames = input();
        this.style = input();
        this.divId = input();
        this.revision = input(0);
        this.className = input();
        this.debug = input(false);
        this.useResizeHandler = input(false);
        this.updateOnLayoutChange = input(true);
        this.updateOnDataChange = input(true);
        this.updateOnlyWithRevision = input(false);
        this.initialized = output();
        this.update = output();
        this.purge = output();
        this.error = output();
        this.afterExport = output();
        this.afterPlot = output();
        this.animated = output();
        this.animatingFrame = output();
        this.animationInterrupted = output();
        this.autoSize = output();
        this.beforeExport = output();
        this.beforeHover = output();
        this.buttonClicked = output();
        /**
         * @deprecated DEPRECATED: Reconsider using `(plotlyClick)` instead of `(click)` to avoid event conflict. Please check https://github.com/plotly/angular-plotly.js#FAQ
         */
        this.click = output();
        this.plotlyClick = output();
        this.clickAnnotation = output();
        this.deselect = output();
        this.doubleClick = output();
        this.framework = output();
        this.hover = output();
        this.legendClick = output();
        this.legendDoubleClick = output();
        /**
         * @deprecated DEPRECATED: Event react is not list as an plotly.js event
         */
        this.react = output();
        this.relayout = output();
        this.relayouting = output();
        this.restyle = output();
        this.redraw = output();
        this.selected = output();
        this.selecting = output();
        this.sliderChange = output();
        this.sliderEnd = output();
        this.sliderStart = output();
        this.sunburstclick = output();
        this.transitioning = output();
        this.transitionInterrupted = output();
        this.unhover = output();
        /**
         * @deprecated DEPRECATED: Event treemapclick is not list as an plotly.js event
         */
        this.treemapclick = output();
        this.webglcontextlost = output();
        this.eventNames = ['afterExport', 'afterPlot', 'animated', 'animatingFrame', 'animationInterrupted', 'autoSize',
            'beforeExport', 'beforeHover', 'buttonClicked', 'clickAnnotation', 'deselect', 'doubleClick', 'framework', 'hover',
            'legendClick', 'legendDoubleClick', 'react', 'relayout', 'relayouting', 'restyle', 'redraw', 'selected', 'selecting', 'sliderChange',
            'sliderEnd', 'sliderStart', 'sunburstclick', 'transitioning', 'transitionInterrupted', 'unhover', 'treemapclick', 'webglcontextlost'];
    }
    ngOnInit() {
        this.createPlot().then(() => {
            const figure = this.createFigure();
            this.initialized.emit(figure);
        });
    }
    ngOnDestroy() {
        if (typeof this.resizeHandler === 'function') {
            this.getWindow().removeEventListener('resize', this.resizeHandler);
            this.resizeHandler = undefined;
        }
        if (this.plotlyInstance) {
            const figure = this.createFigure();
            this.purge.emit(figure);
            PlotlyService.remove(this.plotlyInstance);
        }
    }
    ngOnChanges(changes) {
        let shouldUpdate = false;
        const revision = changes['revision'];
        if (revision && !revision.isFirstChange()) {
            shouldUpdate = true;
        }
        const debug = changes['debug'];
        if (debug && !debug.isFirstChange()) {
            shouldUpdate = true;
        }
        if (shouldUpdate) {
            this.updatePlot();
        }
        this.updateWindowResizeHandler();
    }
    ngDoCheck() {
        if (this.updateOnlyWithRevision()) {
            return false;
        }
        let shouldUpdate = false;
        if (this.updateOnLayoutChange()) {
            if (this.layoutDiffer) {
                const layoutHasDiff = this.layoutDiffer.diff(this.layout());
                if (layoutHasDiff) {
                    shouldUpdate = true;
                }
            }
            else if (this.layout()) {
                this.layoutDiffer = this.keyValueDiffers.find(this.layout()).create();
            }
            else {
                this.layoutDiffer = undefined;
            }
        }
        if (this.updateOnDataChange()) {
            if (this.dataDiffer) {
                const dataHasDiff = this.dataDiffer.diff(this.data());
                if (dataHasDiff) {
                    shouldUpdate = true;
                }
            }
            else if (Array.isArray(this.data())) {
                this.dataDiffer = this.iterableDiffers.find(this.data()).create(this.dataDifferTrackBy);
            }
            else {
                this.dataDiffer = undefined;
            }
        }
        if (shouldUpdate && this.plotlyInstance) {
            this.updatePlot();
        }
    }
    getData() {
        return this.data() ?? [];
    }
    getWindow() {
        return window;
    }
    getClassName() {
        let classes = [this.defaultClassName];
        const className = this.className();
        if (Array.isArray(className)) {
            classes = classes.concat(className);
        }
        else if (className) {
            classes.push(className);
        }
        return classes.join(' ');
    }
    createPlot() {
        return this.plotly.newPlot(this.plotEl.nativeElement, this.getData(), this.layout(), this.config(), this.frames()).then(plotlyInstance => {
            this.plotlyInstance = plotlyInstance;
            this.getWindow().gd = this.debug ? plotlyInstance : undefined;
            this.eventNames.forEach(name => {
                const eventName = `plotly_${name.toLowerCase()}`;
                const event = this[name];
                plotlyInstance.on(eventName, (data) => event.emit(data));
            });
            plotlyInstance.on('plotly_click', (data) => {
                this.plotlyClick.emit(data);
            });
            this.updateWindowResizeHandler();
        }, err => {
            console.error('Error while plotting:', err);
            this.error.emit(err);
        });
    }
    createFigure() {
        const p = this.plotlyInstance;
        const figure = {
            data: p.data,
            layout: p.layout,
            frames: p._transitionData ? p._transitionData._frames : null
        };
        return figure;
    }
    updatePlot() {
        if (!this.plotlyInstance) {
            const error = new Error(`Plotly component wasn't initialized`);
            this.error.emit(error);
            throw error;
        }
        const layout = { ...this.layout() };
        return this.plotly.update(this.plotlyInstance, this.getData(), layout, this.config(), this.frames()).then(() => {
            const figure = this.createFigure();
            this.update.emit(figure);
        }, err => {
            console.error('Error while updating plot:', err);
            this.error.emit(err);
        });
    }
    updateWindowResizeHandler() {
        if (this.useResizeHandler()) {
            if (this.resizeHandler === undefined) {
                this.resizeHandler = () => this.plotly.resize(this.plotlyInstance);
                this.getWindow().addEventListener('resize', this.resizeHandler);
            }
        }
        else {
            if (typeof this.resizeHandler === 'function') {
                this.getWindow().removeEventListener('resize', this.resizeHandler);
                this.resizeHandler = undefined;
            }
        }
    }
    dataDifferTrackBy(_, item) {
        const obj = Object.assign({}, item, { uid: '' });
        return JSON.stringify(obj);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyComponent, deps: [{ token: PlotlyService }, { token: i0.IterableDiffers }, { token: i0.KeyValueDiffers }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "20.0.2", type: PlotlyComponent, isStandalone: true, selector: "plotly-plot", inputs: { data: { classPropertyName: "data", publicName: "data", isSignal: true, isRequired: false, transformFunction: null }, layout: { classPropertyName: "layout", publicName: "layout", isSignal: true, isRequired: false, transformFunction: null }, config: { classPropertyName: "config", publicName: "config", isSignal: true, isRequired: false, transformFunction: null }, frames: { classPropertyName: "frames", publicName: "frames", isSignal: true, isRequired: false, transformFunction: null }, style: { classPropertyName: "style", publicName: "style", isSignal: true, isRequired: false, transformFunction: null }, divId: { classPropertyName: "divId", publicName: "divId", isSignal: true, isRequired: false, transformFunction: null }, revision: { classPropertyName: "revision", publicName: "revision", isSignal: true, isRequired: false, transformFunction: null }, className: { classPropertyName: "className", publicName: "className", isSignal: true, isRequired: false, transformFunction: null }, debug: { classPropertyName: "debug", publicName: "debug", isSignal: true, isRequired: false, transformFunction: null }, useResizeHandler: { classPropertyName: "useResizeHandler", publicName: "useResizeHandler", isSignal: true, isRequired: false, transformFunction: null }, updateOnLayoutChange: { classPropertyName: "updateOnLayoutChange", publicName: "updateOnLayoutChange", isSignal: true, isRequired: false, transformFunction: null }, updateOnDataChange: { classPropertyName: "updateOnDataChange", publicName: "updateOnDataChange", isSignal: true, isRequired: false, transformFunction: null }, updateOnlyWithRevision: { classPropertyName: "updateOnlyWithRevision", publicName: "updateOnlyWithRevision", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { initialized: "initialized", update: "update", purge: "purge", error: "error", afterExport: "afterExport", afterPlot: "afterPlot", animated: "animated", animatingFrame: "animatingFrame", animationInterrupted: "animationInterrupted", autoSize: "autoSize", beforeExport: "beforeExport", beforeHover: "beforeHover", buttonClicked: "buttonClicked", click: "click", plotlyClick: "plotlyClick", clickAnnotation: "clickAnnotation", deselect: "deselect", doubleClick: "doubleClick", framework: "framework", hover: "hover", legendClick: "legendClick", legendDoubleClick: "legendDoubleClick", react: "react", relayout: "relayout", relayouting: "relayouting", restyle: "restyle", redraw: "redraw", selected: "selected", selecting: "selecting", sliderChange: "sliderChange", sliderEnd: "sliderEnd", sliderStart: "sliderStart", sunburstclick: "sunburstclick", transitioning: "transitioning", transitionInterrupted: "transitionInterrupted", unhover: "unhover", treemapclick: "treemapclick", webglcontextlost: "webglcontextlost" }, providers: [PlotlyService], viewQueries: [{ propertyName: "plotEl", first: true, predicate: ["plot"], descendants: true, static: true }], usesOnChanges: true, ngImport: i0, template: `<div #plot [attr.id]="divId()" [ngClass]="getClassName()" [ngStyle]="style()">
      <ng-content></ng-content>
    </div>`, isInline: true, dependencies: [{ kind: "ngmodule", type: CommonModule }, { kind: "directive", type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i2.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'plotly-plot',
                    standalone: true,
                    imports: [CommonModule],
                    template: `<div #plot [attr.id]="divId()" [ngClass]="getClassName()" [ngStyle]="style()">
      <ng-content></ng-content>
    </div>`,
                    providers: [PlotlyService],
                }]
        }], ctorParameters: () => [{ type: PlotlyService }, { type: i0.IterableDiffers }, { type: i0.KeyValueDiffers }], propDecorators: { plotEl: [{
                type: ViewChild,
                args: ['plot', { static: true }]
            }] } });

class PlotlyModule {
    constructor() {
        if (!this.isValid()) {
            const msg = 'Invalid PlotlyJS object. Please check https://github.com/plotly/angular-plotly.js#quick-start'
                + ' to see how to add PlotlyJS to your project.';
            throw new Error(msg);
        }
    }
    isValid() {
        return PlotlyService.plotly !== undefined
            && (typeof PlotlyService.plotly.plot === 'function'
                || typeof PlotlyService.plotly.newPlot === 'function');
    }
    static forRoot(plotlyjs) {
        PlotlyService.setPlotly(plotlyjs);
        return {
            ngModule: PlotlyModule,
            providers: [PlotlyService]
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "20.0.2", ngImport: i0, type: PlotlyModule, imports: [PlotlyComponent], exports: [PlotlyComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyModule, providers: [PlotlyService], imports: [PlotlyComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.2", ngImport: i0, type: PlotlyModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [PlotlyComponent],
                    providers: [PlotlyService],
                    exports: [PlotlyComponent],
                }]
        }], ctorParameters: () => [] });

/*
 * Public API Surface of plotly
 */

/**
 * Generated bundle index. Do not edit.
 */

export { PlotlyComponent, PlotlyModule, PlotlyService };
